File: Estruturais/Adapter/Bridge/Bridge01.js
// Implementação da interface de cor 
class Color{
    constructor(color){
        this.color = color;
    }

    getColor(){
        return this.color;
    }
}

//Imprementações concretas de cores
class RedColor extends Color{
    constructor(){
        super("Red");
    }
}

class BlueColor extends Color{
    constructor(){
        super("Blue");
    }
}

class GreenColor extends Color{
    constructor(){
        super("Green");
    }
}

// implementaçãode interface de forma
class Shape{
    constructor(color){
        this.color = color;
    }
    setColor(color){
        this.color = color;
    }

    draw(){
        throw new Error("Este método precisa ser implementado pela subclasse");
    }
}

// Abstrações Refinidas
class CircleShape extends Shape{
    draw(){
        console.log(`Desenha  um circulo ${this.color.getColor()}.`);

    }
}


class SquareShape extends Shape{
    draw(){
        console.log(`Desenha  um quadrado ${this.color.getColor()}.`);
        
    }
}


// Cliente - Ultilização de método #####
const red = new RedColor();
const blue = new BlueColor();
const green = new GreenColor();



const circle1 = new CircleShape(red);
const circle2 = new CircleShape(green);
const square1 = new SquareShape(blue);
const square2 = new SquareShape(red);


circle1.draw();
circle2.draw();
square1.draw();
square2.draw();

File: Estruturais/Adapter/Bridge/Bridge01.py
# Implementação da interface de cor
class Color:
    def __init__(self, color):
        self.color = color

    def getColor(self):
        return self.color

# Impelmentações concretas de cores
class RedColor(Color):
    def __init__(self):
        super().__init__("Red")

class BlueColor(Color):
    def __init__(self):
        super().__init__("Blue")

class GreenColor(Color):
    def __init__(self):
        super().__init__("Green")

# Implementação da interface de forma
class Shape:
    def __init__(self, color):
        self.color = color
   
    def setColor(self, color):
        self.color = color

    def draw(self):
        raise NotImplementedError("Este método precisa ser implementado pela subclasse")

# Abstrações Refinadas
class CircleShape(Shape):
    def draw(self):
        print(f"Desenhando um circulo {self.color.getColor()}.")

class SquareShape(Shape):
    def draw(self):
        print(f"Desenhando um quadrado {self.color.getColor()}.")

class ShapeFactory:
    def createShape(self, color):
        raise NotImplementedError("Este método precisa ser implementado pela subclasse")

class CircleShapeFactory(ShapeFactory):
    def createShape(self, color):
        return CircleShape(color)

class SquareShapeFactory(ShapeFactory):
    def createShape(self, color):
        return SquareShape(color)

# Cliente - Utilização do método #######################
red = RedColor()
blue = BlueColor()
green = GreenColor()

circleShapeFactory = CircleShapeFactory()
squareShapeFactory = SquareShapeFactory()

circle1 = circleShapeFactory.createShape(red)
circle2 = circleShapeFactory.createShape(green)
square1 = squareShapeFactory.createShape(blue)
square2 = squareShapeFactory.createShape(green)

circle1.draw()
square1.draw()
circle2.draw()
square2.draw()

